pkg/carbon/carbonassets.go

```diff
@@ -0,0 +1,171 @@
+package carbon
+
+import (
+	"embed"
+	"encoding/csv"
+	"strconv"
+	"strings"
+
+	"github.com/opencost/opencost/core/pkg/log"
+	"github.com/opencost/opencost/core/pkg/opencost"
+	"github.com/opencost/opencost/core/pkg/util"
+)
+
+//go:embed carbonlookupdata.csv
+var f embed.FS
+
+var carbonLookupNode map[interface{}]float64
+var carbonLookupDisk map[interface{}]float64
+
+// Opencost does not build network types
+var carbonValidInstanceTypes map[string]string
+var carbonValidRegions map[string]string
+
+func init() {
+
+	carbonData, err := f.ReadFile("carbonlookupdata.csv")
+	if err != nil {
+		log.Errorf("Error getting content of carbon lookup file: %s", err)
+		return
+	}
+
+	reader := csv.NewReader(strings.NewReader(string(carbonData)))
+
+	// skip header
+	_, err = reader.Read()
+	if err != nil {
+		log.Errorf("Error reading carbon lookup data: %s", err)
+		return
+	}
+
+	dat, err := reader.ReadAll()
+	if err != nil {
+		log.Errorf("Error reading carbon lookup data: %s", err)
+		return
+	}
+
+	carbonLookupNode = make(map[interface{}]float64)
+	carbonLookupDisk = make(map[interface{}]float64)
+
+	carbonValidInstanceTypes = make(map[string]string)
+	carbonValidRegions = make(map[string]string)
+
+	for _, carbonItem := range dat {
+
+		if coeff, err := strconv.ParseFloat(carbonItem[5], 64); err != nil {
+
+			log.DedupedWarningf(5, "Error setting carbon lookup item: malformed carbon cost '%s'", carbonItem[5])
```

Comment:
In general, we should avoid the deduped log methods because they dedupe on the format string, not the full error, so we can swallow a ton of errors because of it. Since this is an `init()` I think it is fine to log at Error level without deduping, but I'd personally go further and `panic()` here because `init()` will only be called once, on load.

Reply:
https://stackoverflow.com/questions/24790175/when-is-the-init-function-run